toxcore/group.c

```diff
@@ -2875,22 +2900,242 @@ void send_name_all_groups(Group_Chats *g_c)
+        const Group_Peer *peer = (j < g->numpeers) ? &g->group[j] : &g->frozen[j - g->numpeers];
+
+        if (id_equal(peer->real_pk, g->real_pk)) {
+#ifndef NDEBUG
+            found_self = true;
+#endif
+            continue;
+        }
+
+        data = save_peer(peer, data);
+    }
+
+    assert(found_self);
+
+    return data;
+}
+
+static uint32_t conferences_section_size(const Group_Chats *g_c)
+{
+    uint32_t len = 0;
+
+    for (uint16_t i = 0; i < g_c->num_chats; ++i) {
+        Group_c *g = get_group_c(g_c, i);
+
+        if (!g || g->status != GROUPCHAT_STATUS_CONNECTED) {
+            continue;
+        }
+
+        len += saved_conf_size(g);
+    }
+
+    return len;
+}
+
 uint32_t conferences_size(const Group_Chats *g_c)
 {
-    return 0;
+    return 2 * sizeof(uint32_t) + conferences_section_size(g_c);
 }
 
 uint8_t *conferences_save(const Group_Chats *g_c, uint8_t *data)
 {
+    const uint32_t len = conferences_section_size(g_c);
+    data = state_write_section_header(data, STATE_COOKIE_TYPE, len, STATE_TYPE_CONFERENCES);
+
+    for (uint16_t i = 0; i < g_c->num_chats; ++i) {
+        Group_c *g = get_group_c(g_c, i);
+
+        if (!g || g->status != GROUPCHAT_STATUS_CONNECTED) {
+            continue;
+        }
+
+        data = save_conf(g, data);
+    }
+
     return data;
 }
 
+static State_Load_Status load_conferences(Group_Chats *g_c, const uint8_t *data, uint32_t length)
+{
+    const uint8_t *init_data = data;
+
+    while (length >= (uint32_t)(data - init_data) + SAVED_CONF_SIZE_CONSTANT) {
+        const int groupnumber = create_group_chat(g_c);
+
+        if (groupnumber == -1) {
+            return STATE_LOAD_STATUS_ERROR;
+        }
+
+        Group_c *g = &g_c->chats[groupnumber];
+
+        g->type = *data;
+        ++data;
+
+        memcpy(g->id, data, GROUP_ID_LENGTH);
+        data += GROUP_ID_LENGTH;
+
+        lendian_bytes_to_host32(&g->message_number, data);
+        data += sizeof(uint32_t);
+
+        lendian_bytes_to_host16(&g->lossy_message_number, data);
+        data += sizeof(uint16_t);
+
+        lendian_bytes_to_host16(&g->peer_number, data);
+        data += sizeof(uint16_t);
+
+        lendian_bytes_to_host32(&g->numfrozen, data);
+        data += sizeof(uint32_t);
+
+        g->frozen = (Group_Peer *)malloc(sizeof(Group_Peer) * g->numfrozen);
+
+        if (g->frozen == nullptr) {
+            return STATE_LOAD_STATUS_ERROR;
+        }
+
+        g->title_len = *data;
+        ++data;
+
+        if (length < (uint32_t)(data - init_data) + g->title_len) {
+            return STATE_LOAD_STATUS_ERROR;
+        }
+
+        memcpy(g->title, data, g->title_len);
+        data += g->title_len;
+
+        for (uint32_t j = 0; j < g->numfrozen; ++j) {
+            if (length < (uint32_t)(data - init_data) + SAVED_PEER_SIZE_CONSTANT) {
+                return STATE_LOAD_STATUS_ERROR;
+            }
+
+            Group_Peer *peer = &g->frozen[j];
+            memset(peer, 0, sizeof(Group_Peer));
```

Comment:
This is not ideal. Are we not setting every field of `Group_Peer` explicitly while loading it?

Reply:
Well, this is also how the peer structure is initialised in addpeer. Should I add an initialisation function to be used in both places, explicitly setting the various fields to 0?